lib/linkerd2-proxy-api-resolve/src/resolve.rs

```diff
@@ -0,0 +1,144 @@
+use crate::api::destination as api;
+use crate::core::resolve::Update;
+use crate::metadata::Metadata;
+use crate::pb;
+use futures::{future, Async, Future, Poll, Stream};
+use tower::Service;
+use tower_grpc::{self as grpc, generic::client::GrpcService, Body, BoxBody};
+use tracing::{debug, trace};
+
+#[derive(Clone)]
+pub struct Resolve<S> {
+    service: api::client::Destination<S>,
+    scheme: String,
+    context_token: String,
+}
+
+pub struct Resolution<S: GrpcService<BoxBody>> {
+    inner: grpc::Streaming<api::Update, S::ResponseBody>,
+}
+
+// === impl Resolver ===
+
+impl<S> Resolve<S>
+where
+    S: GrpcService<BoxBody> + Clone + Send + 'static,
+    S::ResponseBody: Send,
+    <S::ResponseBody as Body>::Data: Send,
+    S::Future: Send,
+{
+    pub fn new(svc: S) -> Self {
+        Self {
+            service: api::client::Destination::new(svc),
+            scheme: "".into(),
+            context_token: "".into(),
+        }
+    }
+
+    pub fn with_scheme<T: ToString>(self, scheme: T) -> Self {
+        Self {
+            scheme: scheme.to_string(),
+            ..self
+        }
+    }
+
+    pub fn with_context_token<T: ToString>(self, context_token: T) -> Self {
+        Self {
+            context_token: context_token.to_string(),
+            ..self
+        }
+    }
+}
+
+impl<T, S> Service<T> for Resolve<S>
+where
+    T: ToString,
+    S: GrpcService<BoxBody> + Clone + Send + 'static,
+    S::ResponseBody: Send,
+    <S::ResponseBody as Body>::Data: Send,
+    S::Future: Send,
+{
+    type Response = Resolution<S>;
+    type Error = grpc::Status;
+    type Future = future::Map<
+        grpc::client::server_streaming::ResponseFuture<api::Update, S::Future>,
+        fn(grpc::Response<grpc::Streaming<api::Update, S::ResponseBody>>) -> Resolution<S>,
+    >;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        self.service.poll_ready()
+    }
+
+    fn call(&mut self, target: T) -> Self::Future {
+        let path = target.to_string();
+        trace!("resolve {:?}", path);
+        self.service
+            .get(grpc::Request::new(api::GetDestination {
+                path,
+                scheme: self.scheme.clone(),
+                context_token: self.context_token.clone(),
+            }))
+            .map(|rsp| {
+                debug!(metadata = ?rsp.metadata());
+                Resolution {
+                    inner: rsp.into_inner(),
+                }
+            })
```

Comment:
So, I'm not opposed to it, but this raises a larger point.. When is this necessary?

In this case, functionally, all of this is driven on the `discover::buffer`'s task, which is already instrumented with the target...

Now, on the other hand, api-resolve can be used without `discover::buffer`, so _something_ needs to be able to set this...

In general, I think this points to instrumentation being _never_ being done in arbitrary middlewares and _only_ being initiated through a dedicated tracing middleware.  Other middlewares shouldn't re-log, the target, though they can/should log the fact they are being called.

Reply:
> In this case, functionally, all of this is driven on the discover::buffer's task, which is already instrumented with the target...

ah, yeah, you're right; this would only need a span of its own if we wanted to record the time spent executing _this_ future in particular. i hadn't realised the buffer task was also instrumented because i was reading through the code top to bottom, which is probably a bug on my side :P